### PR TITLE
misc: split database and added docs

### PR DESCRIPTION
**Pull request by baszalmstra**
It required some effort to understand the codebase so I thought it was time to add a bunch of documentation. I also split some of the databases into different databases to get a better grouping.

### CODE REVIEW

File: book/src/ch04-00-developer-docs.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# Salsa
+
+The Mun backend makes extensive use of 
+[Salsa](https://github.com/salsa-rs/salsa) to perform on demand, incremental
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:4]:
```suggestion
[Salsa](https://github.com/salsa-rs/salsa) to perform on-demand, incremental 
```

File: book/src/ch04-00-developer-docs.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
+# Salsa
+
+The Mun backend makes extensive use of 
+[Salsa](https://github.com/salsa-rs/salsa) to perform on demand, incremental 
+computations. `salsa` lets you define your program as a set of *queries*. Every 
+query is used like a function `K -> V` that maps from some key `K` to a value of
+`V`. Changing the input `K` also invalidates the result of computing `K -> V`. 
+However, if the result `V` does not change after changing `K` the invalidation 
+is not propagated to computations that use `V` as an input. This enables Mun to
+cache a lot of computations between compilations enabling
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:10]:
```suggestion
cache a lot of computations between compilations, enabling 
```

File: book/src/ch04-00-developer-docs.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+cache a lot of computations between compilations enabling 
+very fast rebuilding on incremental changes.
+
+> For more in depth information on Salsa please refer to the 
+>[Salsa repository](https://github.com/salsa-rs/salsa) or the
+>[Salsa book](https://salsa-rs.github.io/salsa/)
+
+## Databases
+
+Queries are grouped together in so called *databases* defined as traits.
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:19]:
```suggestion
Queries are grouped together in so-called *databases* defined as traits. 
```

File: crates/mun_hir/src/db.rs [hunk 4]
```diff
@@ -53,13 +48,34 @@ pub trait SourceDatabase: salsa::Database {
...
+    /// Parses the file into the syntax tree.
+    #[salsa::invoke(parse_query)]
+    fn parse(&self, file_id: FileId) -> Parse<ast::SourceFile>;
+
     /// Returns the top level AST items of a file
     #[salsa::invoke(crate::source_id::AstIdMap::ast_id_map_query)]
     fn ast_id_map(&self, file_id: FileId) -> Arc<AstIdMap>;
+}
+
+/// The `InternDatabase` maps certain datastructures to ids. These ids refer to instances of
```
Comment (Wodann) [inline crates/mun_hir/src/db.rs:64]:
```suggestion
/// The `InternDatabase` maps certain data structures to ids. These ids refer to instances of
```

File: book/src/ch04-01-salsa.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Salsa
+
+The Mun backend makes extensive use of [Salsa](https://github.com/salsa-rs/salsa) to perform on demand, incremental computations. `salsa` lets you define your program as a set of *queries*. Every query is used like a function `K -> V` that maps from some key `K` to a value of `V`. Changing the input `K` also invalidates the result of computing `K -> V`. However, if the result `V` does not change after changing `K` the invalidation is not propagated to computations that use `V` as an input.
```
Comment (Wodann) [inline book/src/ch04-01-salsa.md:3]: This is the same as ch04-00

File: book/src/ch04-00-developer-docs.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+>[Rust Analyzer](https://github.com/rust-analyzer/rust-analyzer).
+
+### `SourceDatabase`
+
+The `SourceDatabase` provides queries and inputs that form the basis of all
+ compilation because it contains the original source text. 
+
+The `SourceDatabase` holds `SourceRoot`s which are groupings of files. 
+`SourceRoot`s define the hierarchy of files relative to a certain root. For the 
+system its not interesting where the files come from and how they are actually
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:37]:
```suggestion
system it's not interesting where the files come from and how they are actually 
```

File: book/src/ch04-00-developer-docs.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+The `SourceDatabase` holds `SourceRoot`s which are groupings of files. 
+`SourceRoot`s define the hierarchy of files relative to a certain root. For the 
+system its not interesting where the files come from and how they are actually 
+organized on disk. `SourceRoot`s provide an abstraction over the filesystem that
+ provides just enough information for the backend to function.
+
+Another relevant query is the `line_index` query which provides line and column 
+information for characters in a source file. This is used for instance when 
+emitting human readable diagnostic messages with line and column numbers.
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:43]:
```suggestion
emitting human-readable diagnostic messages with line and column numbers.
```

File: book/src/ch04-00-developer-docs.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+### `AstDatabase`
+
+The `AstDatabase` provides Syntax Tree transformations on top of the source
+ text. The `parse` returns the
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:48]:
```suggestion
 text. The `parse` function returns the 
```

File: book/src/ch04-00-developer-docs.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+Another relevant query is the `line_index` query which provides line and column 
+information for characters in a source file. This is used for instance when 
+emitting human readable diagnostic messages with line and column numbers.
+
+### `AstDatabase`
+
+The `AstDatabase` provides Syntax Tree transformations on top of the source
+ text. The `parse` returns the 
+[Concrete Syntax Tree](https://en.wikipedia.org/wiki/Parse_tree) (CST) provided
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:49]: This sentence seems incorrect, but I'm not sure what you are trying to express

File: book/src/ch04-00-developer-docs.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+ text. The `parse` returns the 
+[Concrete Syntax Tree](https://en.wikipedia.org/wiki/Parse_tree) (CST) provided 
+of a file by the `mun_syntax` crate. Any errors related to parsing are stored in
+ the tree itself. 
+
+As an example of where Salsa helps Mun achieve faster incremental compilation: 
+changing a single character in a file will also change the CST which in turn
+will invalidate a lot of computations. However, most changes occur within blocks 
+(like functions or structs). To not invalidate everything in a file on a
+change another query transforms the CST to an `AstIdMap` which assigns ids to
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:57]:
```suggestion
(like functions or structs). To not invalidate everything in the changed file,
another query transforms the CST to an `AstIdMap` which assigns ids to
```

File: book/src/ch04-00-developer-docs.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+all top level items in a file. This enables referring to those ids in later
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:58]:
```suggestion
all top-level items in a file. This enables referring to those ids in later
```

File: book/src/ch04-00-developer-docs.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+stages of compilation which will change much less often.
+
+### `InternDatabase`
+
+The `InternDatase` links certain locations in the Syntax Tree to ids. For
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:63]:
```suggestion
The `InternDatase` links certain locations in the syntax tree to ids. For
```

File: book/src/ch04-00-developer-docs.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+instance it associates a function syntax node with a `FunctionId`. This id
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:64]:
```suggestion
instance, it associates a function syntax node with a `FunctionId`. This id
```

File: book/src/ch04-00-developer-docs.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+is then used everywhere in code to refer to the specific function instance. The 
+same happens for other items like structs. The purpose of this *interning* is 
+again to allow better incremental compilation for queries that only need to 
+refer to certain constructs without having to refer to the volatile syntax tree
+elements. 
+
+### `DefDatabase`
+
+The `DefDatabase` provides definitions extracted from the syntax trees. It 
+provides desugered or lowered versions of language constructs that enable other
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:74]:
```suggestion
provides desugared or lowered versions of language constructs that enable other
```

File: book/src/ch04-00-developer-docs.md [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+queries to query specific information for these constructs. Extracting this data
+again enables faster incremental compilation because if the extracted data does 
+not change dependant queries also will not change.
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:77]:
```suggestion
not change dependent queries also will not change.
```

File: book/src/ch04-00-developer-docs.md [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+### `HirDatabase`
+
+The `HirDatabase` resolves all types and names from the definitions. It also 
+performs type inferencing. The resulting High Level Representation (HIR) is the
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:82]:
```suggestion
performs type inferencing. The resulting High-Level Representation (HIR) is the
```

File: book/src/ch04-00-developer-docs.md [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+input for the LLVM Intermediate Representation (LLVM IR). The HIR datastructures
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:83]:
```suggestion
input for the LLVM Intermediate Representation (LLVM IR). The HIR data structures
```

File: book/src/ch04-00-developer-docs.md [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+can also contain errors only if there are no errors present will generating IR
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:84]:
```suggestion
can also contain errors, but only if there are no errors present, will generating IR
```

File: book/src/ch04-00-developer-docs.md [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+Finally, the `CodeGenDatabase` provides queries to construct assemblies from 
+groups of files. Assemblies are the final product of the Mun compiler. They are 
+the binaries that can be executed by the runtime. This stage is where Salsa 
+really shines; only if something actually changed in the HIR will an assembly be 
+created. 
+
+## Language Server
+
+The language server uses the same backend as the compiler. Where the compiler
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:97]:
```suggestion
The language server uses the same backend as the compiler. Whereas the compiler 
```

File: book/src/ch04-00-developer-docs.md [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+Finally, the `CodeGenDatabase` provides queries to construct assemblies from 
+groups of files. Assemblies are the final product of the Mun compiler. They are 
+the binaries that can be executed by the runtime. This stage is where Salsa 
+really shines; only if something actually changed in the HIR will an assembly be 
+created. 
+
+## Language Server
+
+The language server uses the same backend as the compiler. Where the compiler 
+creates assemblies the language server queries the different database to provide
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:98]:
```suggestion
creates assemblies, the language server queries the different databases to provide
```

File: book/src/ch04-00-developer-docs.md [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+a user with code intelligence. It also uses the exact same diagnostics paths as 
+the compiler which means there is never a mismatch between the compiler and the
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:100]:
```suggestion
the compiler, which means there is never a mismatch between the compiler and the 
```

File: book/src/ch04-00-developer-docs.md [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+really shines; only if something actually changed in the HIR will an assembly be 
+created. 
+
+## Language Server
+
+The language server uses the same backend as the compiler. Where the compiler 
+creates assemblies the language server queries the different database to provide
+a user with code intelligence. It also uses the exact same diagnostics paths as 
+the compiler which means there is never a mismatch between the compiler and the 
+language server and vice versa. If compiling your code will result in an error
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:101]:
```suggestion
language server - and vice versa. If compiling your code results in an error, 
```

File: book/src/ch04-00-developer-docs.md [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+created. 
+
+## Language Server
+
+The language server uses the same backend as the compiler. Where the compiler 
+creates assemblies the language server queries the different database to provide
+a user with code intelligence. It also uses the exact same diagnostics paths as 
+the compiler which means there is never a mismatch between the compiler and the 
+language server and vice versa. If compiling your code will result in an error 
+it will always also be immediately visible in your IDE.
```
Comment (Wodann) [inline book/src/ch04-00-developer-docs.md:102]:
```suggestion
it will also be immediately visible in your IDE. 
```

**Review by Wodann — CHANGES_REQUESTED** [review]
Good to have this documentation, and the code is also a lot more concise. Thanks for this!

How did you make the diagram? I like the style! 🙂

File: book/src/ch04-01-salsa.md (hunk 5 above)
Comment (baszalmstra) [reply book/src/ch04-01-salsa.md:3]: Thats a mistake, I changed that

**Comment by baszalmstra** [comment]
> How did you make the diagram? I like the style! 🙂

It's basically draw.io with a custom font: https://drawio-app.com/introducing-draw-ios-new-sketch-feature/
The font is: https://fonts.google.com/specimen/Indie+Flower

**Review by Wodann — APPROVED** [review]

**Pull request merged**